internal/controller/promotion/argocd.go

```diff
@@ -156,6 +157,16 @@ func (a *argoCDMechanism) Promote(
 				logger.Info(err.Error())
 			}
 			if phase.Failed() {
+				// Record the reason for the failure if available.
+				if app.Status.OperationState != nil {
```

Comment:
This piggybacks on the fact that we short-circuit. In the initial version, I collected the messages together with the phases. But this grew into something unnecessarily complex and over-engineered, which I would like to avoid as long as we do not need it.